### PR TITLE
Store: store structured data as JSON objects

### LINKED ISSUE #19: Invalid JSON
When dumping the DataModel `m_vars` after tools have added their own config to it via `ExportConfiguration()`, invalid JSON is produced by the `Store::operator>>()` function

e.g.
```
{"$0":"./main" ,"$1":"WCSim_NHitsOnly_FlatOut_1ThreadPerTool" ,"in_wcsim":"{"verbose":"3" }"  }
```
The offending quotes basically what's happening is that the sub-block within `{...}` is quoted like `"{...}"`

### PR DESCRIPTION
**Pull request by jini-zh**
This pull request makes an overhaul of `Store` implementation, extending its support of edge cases in user data.

The primary change is that the structured data stored in a `Store` such as vectors and maps, are stored as JSON arrays and objects, with the strings in keys and values properly encoded, so a `"q } w"` should result the same string when requested, and it won't confuse the JSON parser and serializer.

The type of an object stored is encoded by the first character of its serialized string: `"` means that it's a string, `[` --- a vector, `{` --- a map. Any other character means that it is an object with custom encoding (see below). Strings and custom objects are stored as is, but when serialized to JSON format, their contents are properly escaped to produce a valid JSON string. Vectors and maps are stored in JSON format initially.

The user can add support for storage of their own classes `T` by implementing `operator<<(std::ostream&, T)` and `operator>>(std::istream&, T&)`, or `json_encode_r(std::ostream&, T)` and `json_decode_r(const char*&, T&)`. When both the operators and the functions are defined, the functions take priority. `json_encode_r` and `json_decode_r` extend the JSON parser and serializer to support custom user classes and are useful outside the `Store` class.

Speaking of JSON, this pull request also adds generic functions for JSON production, validation, and parsing, see `json_encode`, `json_scan*` and `json_decode` in `Json.h`.

`Store::Initialise` has been updated to allow escaped comments, newlines, and spaces at the end of a line, and to keep all of the spaces in the value string, so

    var long   space \# test \\ \
      newline  \ 

results in `"long   space # test \\ \n newline   "`. This change is not backwards compatible, but it is unlikely that anyone relied on that feature.

`UnitTests/StoreTest` has been rewritten to test for many edge cases and the new features.

I personally think that we should stop abusing `Store` for JSON data and instead either implement a custom `JSON` object that stores values as a tree in memory, or just use the new `json_encode` and `json_decode` functions for (de-)serialization.

I have checked that libDAQInterface Example works with the new `Store`, but I think is worth testing this code on some real codebase before merging this pull request. Can someone do that?

This pull request fixes #19, #22, #25. #16 appears to be not relevant anymore.

### CODE REVIEW

File: src/Store/Json.cpp [hunk 1]
```diff
@@ -1,19 +1,387 @@
...
+  return json_encode(output, datum.begin(), datum.end());
+}
+
+bool json_encode_r(std::ostream& output, const char* datum, adl_tag) {
+  return json_encode(output, datum, datum + strlen(datum));
+}
+
+const char* json_scan_whitespace(const char* input) {
+  if (!input) return nullptr;
+  while (isspace(*input)) ++input;
```
Comment (marc1uk) [inline src/Store/Json.cpp:80]: possible segfault if the input does not contain a whitespace character, it'll just keep going, potentially out of the user's input buffer, and until it tries to deref a nullptr.

Comment (jini-zh) [reply src/Store/Json.cpp:80]: If there are no whitespace characters in `input`, `isspace` will return `false` and the iteration in line 80 will stop immediately

Comment (marc1uk) [reply src/Store/Json.cpp:80]: ah my bad, i should have said if it only contains whitespace, which admittedly is much less likely.

Comment (jini-zh) [reply src/Store/Json.cpp:80]: `isspace(0)` is false, so the iteration won't go beyond the end of the string.

**Review by marc1uk — APPROVED** [review]
